### PR TITLE
Add autotype

### PR DESCRIPTION
**Pull request by innovate-invent**
Some work still needs to be done to implement a popover model window to select a target when multiple matches occur. There are also other // TODO items in the code that are not critical.

I haven't looked at the changes needed to the server to store the new autotype rules.

Provides Keepass TCATO functionality, except rather than using a seeded rand, it compares to the median letter in the string. Anything less than that letter is typed using the clipboard. The consequence is weak, repetitive passwords stay weak.

Tries to be compatible with keepass placeholders.
{ALT}, {CTRL}, {SHIFT[ RIGHT|LEFT]} are added to allow tapping those buttons.

Some features waiting on https://github.com/octalmage/robotjs/pull/357 and https://github.com/octalmage/robotjs/issues/490

### CODE REVIEW

**Comment by kspearrin** [comment]
Thanks for the PR. I will look into this more in the future when I have some time.

**Comment by krestaino** [comment]
Any update here? This would be a really nice addition to the desktop app.

**Comment by innovate-invent** [comment]
Looks like the missing modal was already implemented in https://github.com/bitwarden/desktop/pull/231

**Comment by immanuelfodor** [comment]
The most killer feature od KeePass is the autotype (especially the two-channel autotype), and this was truly missing from Bitwarden until now. I hope you can make this feature to the master branch! :)

**Comment by utahbmxer** [comment]
I would love this to PR to be completed as well.  I tried to download the remote branch and build it, but I have never built NodeJS stuff and got lost.  :(

**Comment by immanuelfodor** [comment]
There are also 5 conflicts, so the PR needs to be updated by the author @innovate-invent. 
When you say some features are waiting in external dependencies, what do you mean? If the PR would be merged as is, what would be missing/unavailable?

**Comment by innovate-invent** [comment]
I have actually abandoned bitwarden. Have a look at the linked issues for more information about the dependencies needed for some features.

File: src/services/auto-type.service.ts [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+    }
+
+    typeUsingClip(str: string) {
+        if (str.length === 0) { return; }
+        this.platformUtilsService.copyToClipboard(str);
+        robot.keyTap('v', this.systemClipModifier);
+    }
+
+    async typeTarget() {
+        const target = await this.getTarget();
```
Comment (innovate-invent) [inline src/services/auto-type.service.ts]: Looks like I missed something while writing this. typeTarget() needs to take the target as a parameter. Line 237 needs to be changed to `typeTarget(windowManager.getActiveWindow())`.

File: src/services/auto-type.service.ts [hunk 2]
```diff
@@ -0,0 +1,936 @@
...
+            }
+        }
+
+        // Decide on final candidate
+        let candidate: {match: RegExpMatchArray, target: AutoTypeView, cipher: Cipher, decCipher: CipherView} = null;
+        if (candidates.length === 1) {
+            candidate = candidates[0];
+        } else if (candidates.length > 1) {
+            candidates.sort((a, b) => b.match.length - a.match.length);
+            candidate = candidates[0]; // TODO show context menu to pick candidate
```
Comment (innovate-invent) [inline src/services/auto-type.service.ts]: This line needs to be replaced with the candidate selection modal that still has to be written.

**Comment by andrew-oreshko** [comment]
Do you guys consider to merge it anytime soon?

**Comment by cscharf** [comment]
Hi All,

We certainly do want to consider adding AutoType and have this on our radar, however we are looking at a holistic approach to engineering this feature. In that regard this PR can't be used for a few reasons since it is relatively incomplete, it relies on other 3rd party packages that have known issues/incomplete support, and the original submitter has not signed our CLA. We may be able to use some of this conceptually but that has a bit to be seen once we're able to look at this more holistically. With that, we're going to keep AutoType strong on our feature request list and will certainly invest some engineering fire-power towards this in the not-too-distant future, however I'm closing this PR at this time.

**Comment by MCOfficer** [comment]
Since this was closed, is there a tracking issue we can subscribe to instead?

**Comment by tgreer-bw** [comment]
Not a live-issue per se, but when we're actively developing we'll update the roadmap here:
https://community.bitwarden.com/t/bitwarden-roadmap/

And the thread here:
https://community.bitwarden.com/t/auto-type-autofill-for-logging-into-other-desktop-apps/

**Comment by ZeroMaki** [comment]
> I have actually abandoned bitwarden. Have a look at the linked issues for more information about the dependencies needed for some features. Everything has been documented, just search src/services/auto-type.service.ts for '//TODO'.

what did you swap to instead? some programs outright BLOCK pasting password so auto type is quite necessary if you're using generated passwords

**Comment by innovate-invent** [comment]
GoPass is what I landed on

**Comment by niikoo** [comment]
+1

**Pull request closed without merging**